src/lib/storage/table.cpp

```diff
@@ -78,13 +93,62 @@ const std::string& Table::column_name(const ColumnID column_id) const { return _
 
 const std::string& Table::column_type(const ColumnID column_id) const { return _column_types.at(column_id); }
 
-std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) { return _chunks.at(chunk_id); }
+std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) {
+  auto guard = std::lock_guard<std::mutex>(_chunk_access_mutex);
+  return _chunks.at(chunk_id);
+}
 
-std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const { return _chunks.at(chunk_id); }
+std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
+  auto guard = std::lock_guard<std::mutex>(_chunk_access_mutex);
+  return _chunks.at(chunk_id);
+}
 
 void Table::compress_chunk(const ChunkID chunk_id) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  auto thread_pool = std::vector<std::thread>{_column_types.size()};
+  const auto& raw_chunk = get_chunk(chunk_id);  // get_chunk performs range check, so we are safe
+  auto compressed_chunk = std::make_shared<Chunk>();
+  auto compressed_segments = std::vector<std::shared_ptr<AbstractSegment>>{_column_types.size()};
+  auto segments_vec_mutex = std::mutex{};
+
+  auto compress_segment = [&raw_chunk, &compressed_segments, &segments_vec_mutex](const auto segment_type,
+                                                                                  const auto column_index) {
+    resolve_data_type(segment_type, [&](const auto data_type_t) {
+      // figure out the type of the segment
+      using ColumnDataType = typename decltype(data_type_t)::type;
+      auto guard = std::lock_guard<std::mutex>(segments_vec_mutex);
```

Comment:
I don't see why another mutex is needed here. You are fetching data with a column_index and setting the result at the position column_index again.

Reply:
We read up and found that std::vector is safe for one concurrent write access or many reads.
Since this function will run on the individual threads, there can potentially be multiple write attempts to ``compressed_segments``, which would be bad, regardless of whether we try to write to different positions, at least from our understanding.